source/agora/consensus/Reward.d

```diff
@@ -0,0 +1,413 @@
+    /// distributed to all the Validators over about 128 years
+    private const Amount totalValRewards = 2_700_000_000.coins;
+
+    /// Every 5 secs 50 coins are rewarded equals 10 per second (Block interval of 600 secs)
+    private const Amount commonsCoinsPerBlock = 6000.coins;
+
+    /// This is the Amount to pay the Commons Budget in a payout period before the allocation runs out
+    private Amount commonsPerPayout;
+
+    /// Number of blocks that rewards are paid for each payout (also the interval for payments)
+    private uint payout_period;
+
+    /// Duration of a block in seconds
+    private ulong block_interval_sec;
+
+    /// Number of Coinbase transactions per year
+    private ulong payouts_per_year;
+
+    /***************************************************************************
+
+        Construct `Reward` object which encapsulates block reward calculations.
+
+        Params:
+            payout_period = how many blocks are included for a reward payout
+            block_interval_sec = time span for each block
+
+    ***************************************************************************/
+
+    public this (in uint payout_period, in Duration block_interval)
+    {
+        this.log = Logger(__MODULE__);
+
+        this.payout_period = payout_period;
+        this.block_interval_sec = block_interval.total!"seconds";
+
+        // Calculate payouts per year
+        this.payouts_per_year = yearOfSecs / (this.payout_period * this.block_interval_sec);
+
+        // Calculate amount that would be paid to Commons Budget for full payout period
+        // if the allocation is still sufficient
+        this.commonsPerPayout = commonsCoinsPerBlock; commonsPerPayout.mul(this.payout_period);
+    }
+
+    /***************************************************************************
+
+        Returns the Block Rewards for the given height.
+
+        Params:
+            height = last block of the current payout period
+                (the CoinBase transaction will be in block at this height plus the payout period)
+            percent_signed = (#actual signatures / #expected signatures) * 100
+                on all the blocks for payout period
+
+        Returns:
+            `BlockRewards` which indicates the `Amount` to pay the `Validators`
+            and the `Commons Budget`
+
+    ***************************************************************************/
+
+    public BlockRewards calculateBlockRewards (in Height height, in ubyte percent_signed) nothrow @safe
+    {
+        assert(height % this.payout_period == 0, "This is not a payout block height");
+        assert (height > 0, "We can not payout in Genesis block");
+
+        auto validator_allocated = allocatedValRewards(height);
+        auto validator = reducedValRewards(validator_allocated, percent_signed);
+        auto leftover = validator_allocated; leftover.sub(validator);
+        auto commons = commonsBudgetRewards(height); commons.add(leftover);
```

Comment:
Can we avoid that pattern of dual statement on the same line ?

When it was first introduced it kinda made sense, because we didn't have operator overloads, and that was only used in test code IIRC. But not checking the return value of `add`, `sub`, and co completely defeats the purpose of `Amount`. And having it hidden on the same line only makes it worse.

Reply:
Updated